### PR TITLE
#28 Geolocation

### PR DESCRIPTION
**Pull request by tayanefernandes**
Story #28

### CODE REVIEW

File: www/js/controller/home.controller.js [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
   $scope.addresses;
 
+  var posOptions = {timeout: 10000, enableHighAccuracy: false};
+
+  $cordovaGeolocation
+    .getCurrentPosition(posOptions)
+    .then(function (position) {
+      $scope.latitude  = position.coords.latitude
+      $scope.longitude = position.coords.longitude
+    }, function(err) {
```
Comment (lucaslago) [inline www/js/controller/home.controller.js]: Essa função sem corpo é necessária, já que não estamos tratando erros?

File: www/test/unit/controller/home.controller.test.js [hunk 2]
```diff
@@ -110,4 +119,20 @@ describe('HomeController', function() {
     expect(scope.search.text).toBe('Av. Ipiranga, 123 - Porto Alegre');
   });
 
+  it('should get the current position', function () {
+    spyOn(cordovaGeolocation, 'getCurrentPosition').and.returnValue(deferred.promise);
```
Comment (lucaslago) [inline www/test/unit/controller/home.controller.test.js]: O que tu acha de testarmos se essa função foi chamada com os parâmetros esperados?

**Comment by tayanefernandes** [comment]
@roberto and @pedroaugustofr please take a look.

File: www/js/controller/home.controller.js [hunk 3]
```diff
@@ -1,10 +1,35 @@
...
+
+  $ionicLoading.show();
+
+  var posOptions = {timeout: 5000, enableHighAccuracy: false};
+
+  $cordovaGeolocation
+    .getCurrentPosition(posOptions)
+    .then(function (position) {
+      $scope.coordinates.latitude  = position.coords.latitude
+      $scope.coordinates.longitude = position.coords.longitude
```
Comment (pedroaugustofr) [inline www/js/controller/home.controller.js]: We don't need to use $scope here, right?

File: www/js/controller/home.controller.js [hunk 4]
```diff
@@ -4,24 +4,22 @@ angular.module('vamosJuntas').controller('HomeController',
...
   $cordovaGeolocation
     .getCurrentPosition(posOptions)
     .then(function (position) {
-      $scope.coordinates.latitude  = position.coords.latitude
-      $scope.coordinates.longitude = position.coords.longitude
 
-      placeFactory.fetchPlaces($scope.coordinates.latitude, $scope.coordinates.longitude).then(function(response) {
+      placeFactory.fetchPlaces(position.coords.latitude, position.coords.longitude).then(function(response) {
         $scope.places = response.data;
       });
```
Comment (tayanefernandes) [inline www/js/controller/home.controller.js:20]: @pedroaugustofr I removed the variable from the scope

File: www/conf/stubserver.js [hunk 5]
```diff
@@ -12,7 +12,7 @@ app.get('/', function (req, res) {
   res.send("hello world");
 });
 
-app.get('/places', function(req, res){
+app.get('/risks-around', function(req, res){
```
Comment (taniadgv) [inline www/conf/stubserver.js]: add space before brackets

File: www/js/controller/home.controller.js [hunk 6]
```diff
@@ -1,10 +1,33 @@
...
+  $scope.errorMessage = false;
+
+  $ionicLoading.show();
+
+  var coordinates = {};
+  var posOptions = {timeout: 5000, enableHighAccuracy: false};
+
+  $cordovaGeolocation
+    .getCurrentPosition(posOptions)
+    .then(function (position) {
```
Comment (taniadgv) [inline www/js/controller/home.controller.js:15]: can we move the $codovaGeolocation.getCurrentPosition() to a factory? usually the controller should not know the business logic.

File: www/js/factory/address.factory.js [hunk 7]
```diff
@@ -17,5 +17,23 @@ angular.module('vamosJuntas').factory('addressFactory', function($http, PlacesAP
...
     },
+
+    getAddressByCoord: function(latitude, longitude) {
+      var deferred = $q.defer();
+      var geocoder = new google.maps.Geocoder();
+      var latlng = {
+          lat: latitude,
+          lng: longitude
+        };
+
```
Comment (taniadgv) [inline www/js/factory/address.factory.js:24]: fix indentation please, I think we are using 2 spaces right?

File: www/js/factory/address.factory.js [hunk 8]
```diff
@@ -17,5 +17,23 @@ angular.module('vamosJuntas').factory('addressFactory', function($http, PlacesAP
...
+      geocoder.geocode({'location': latlng}, function(results, status) {
+        if (status === google.maps.GeocoderStatus.OK && results.length > 0) {
+          deferred.resolve(results[0].formatted_address);
+        } else {
+          deferred.reject(status);
+        }
+      });
+      return deferred.promise;
+    }
```
Comment (taniadgv) [inline www/js/factory/address.factory.js:33]: we should be using the response of the request in the same factory, no need to send it back to the controller's scope.

File: www/test/unit/controller/home.controller.test.js [hunk 9]
```diff
@@ -1,25 +1,36 @@
...
+
+        createController = function() {
+          $controller('HomeController', {
+              '$scope': scope,
+              'placeFactory': placeFactory,
+              'addressFactory': addressFactory,
+              '$cordovaGeolocation': cordovaGeolocation,
+              '$ionicLoading': ionicLoading
+          });
+        };
```
Comment (taniadgv) [inline www/test/unit/controller/home.controller.test.js:26]: please fix indentation

File: www/test/unit/controller/home.controller.test.js [hunk 10]
```diff
@@ -54,20 +65,26 @@ describe('HomeController', function() {
     spyOn(placeFactory, 'fetchPlaces').and.callFake(function() {
       return {
         then: function(callback) {
-          return callback(place);
-        }
-      };
+            return callback(place);
+          }
+        };
```
Comment (taniadgv) [inline www/test/unit/controller/home.controller.test.js]: please fix indentation

File: www/test/unit/controller/home.controller.test.js [hunk 11]
```diff
@@ -114,4 +131,59 @@ describe('HomeController', function() {
...
+      createController();
+      scope.$apply();
+      expect(addressFactory.getAddressByCoord).toHaveBeenCalledWith(-30.057977, -51.1755227);
+      expect(scope.search.text).toBe('Av. Ipiranga, 6681');
+    });
+  });
+
+
+  describe('fails to get current position', function (){
+    beforeEach(function () {
```
Comment (taniadgv) [inline www/test/unit/controller/home.controller.test.js:121]: add space before brackets

**Pull request merged**